thavalon-webapp/src/utils/GameSocket.tsx

```diff
@@ -0,0 +1,61 @@
+interface PingResponse {
+    result: boolean,
+    errorMessage: string, // only populated if result is false
+}
+
+/**
+ * The class containing the underlying websocket for playing the game.
+ * Will contain a websocket, that can either be retrieved or optionally made.
+ */
+export class GameSocket {
+    // the underlying gamesocket instance
+    private static instance: GameSocket;
+    // the instantiated instance of the underlying websocket
+    private websocket: WebSocket;
+
+    // create the underlying websocket in the constructor
+    private constructor(socketUrl: string) {
+        this.websocket = new WebSocket(socketUrl);
+        this.websocket.onopen = this.socketOnOpen;
+        this.websocket.onmessage = this.socketOnMessage;
+        this.websocket.onclose = this.socketOnClose
+        this.websocket.onerror = this.socketOnError;
+    } 
+
+    // functions for handling incoming websocket events
+    private socketOnOpen(event: Event) {
+        console.log("Successfully initiated connection.");
+    }
+
+    private socketOnMessage(event: MessageEvent) {
+        console.log(event);
+        console.log("Received message: " + event.data);
+    }
+
+    private socketOnClose(event: CloseEvent) {
+        console.log("Recieved on close message.");
+    }
+
+    private socketOnError(event: Event) {
+        console.log("Received on error message.");
+    }
+
+    public static getOrCreateInstance(socketUrl?: string): GameSocket {
+        if (socketUrl !== undefined) {
+            GameSocket.instance = new GameSocket(socketUrl);
+        }
+        return GameSocket.instance;

```

Comment:
For @bnavetta, do you think this should split into two functions? My thought right now is we have one function doing two different tasks. If the socketUrl is provided, it just gets the instance. If it's not, it creates an entirely new websocket. Jared's thought is that splitting into a getInstance and a createConnection requires more function calls, since createConnection must be called before the class is useful.

Reply:
I agree with having two functions for two purposes. There'll probably be only 1-2 places where we both create and get a socket, then a bunch where we just get the socket so it's not that many extra function calls. We might also want `createConnection` to close the existing websocket if there is one, and `getInstance` should throw an exception if there's no instance yet.

Reply:
I think if I want it to close the existing weboscket, I need to make it static, since getInstance and createInstance will both be static, and I need to access the websocket to close it.

Reply:
I'm fine with making it static if that's what is needed.